### PR TITLE
nes improvements, mostly softlist related

### PR DESCRIPTION
**Pull request by kazblox**
Mostly from the commit description:
nes.xml: New dumps:
Adan y Eva [hashes were known, but ROMs surfaced on No-Intro]
AV Super Real Pachinko [Frank Cifaldi]

nes.xml: Marked edu2k48 as working, added notes about mouse titles.
Made koko not use battery-backed WRAM as evidenced on PCB picture. (nw)

There's an extra note about GLK Book 48-in-1 if anyone wants to check and disassemble the WORD program to see how it's saving things into the WRAM...

bus/nes/waixing.cpp: Slight modernizations. waixing_sgzlz_device is now
implemented the way CaH4e3 does it, also gets some titles mostly working
in full. (nw)

Note: while the Waixing games under waixing_sgzlz boot up fine under the new implementation of the board, can someone with good knowledge of the Chinese language check if cwdjz works properly? The game seems to boot up past the title screen, but I can't check any further.

fc_suborkey_device: Make PgUp and PgDown keys actually work (nw)

### CODE REVIEW

File: src/devices/bus/nes/waixing.cpp [hunk 1]
```diff
@@ -304,7 +304,7 @@ void nes_waixing_sgzlz_device::pcb_reset()
 	prg32(0);
 	chr8(0, m_chr_source);
 
-	m_latch = 0;
+	memset(m_reg, 0, sizeof(m_reg));
```
Comment (cuavas) [inline src/devices/bus/nes/waixing.cpp]: Please use `std::fill` over memset.

File: src/devices/bus/nes/waixing.cpp [hunk 2]
```diff
@@ -1035,30 +1035,69 @@ WRITE8_MEMBER(nes_waixing_sgz_device::write_h)
...
-			m_latch = (m_latch & 0x0c) | ((data >> 1) & 0x03);
-			prg32(m_latch);
-			break;
-		case 0x702:
-			m_latch = (m_latch & 0x03) | ((data << 2) & 0x0c);
-			break;
+		m_reg[offset & 0x03] = data;
+		uint8_t hbank = m_reg[1] & 0x7;
+		uint8_t lbank = m_reg[2];
+		uint8_t bank = (lbank << 3) | hbank;
```
Comment (cuavas) [inline src/devices/bus/nes/waixing.cpp]: if these locals aren't supposed to be changed after being set, make them const.

File: src/devices/bus/nes/waixing.cpp [hunk 3]
```diff
@@ -142,7 +142,7 @@ nes_waixing_sgz_device::nes_waixing_sgz_device(const machine_config &mconfig, co
 }
 
 nes_waixing_sgzlz_device::nes_waixing_sgzlz_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
-					: nes_nrom_device(mconfig, NES_WAIXING_SGZLZ, "NES Cart Waixing San Guo Zhong Lie Zhuan PCB", tag, owner, clock, "nes_waixing_sgzlz", __FILE__), m_latch(0)
+					: nes_nrom_device(mconfig, NES_WAIXING_SGZLZ, "NES Cart Waixing San Guo Zhong Lie Zhuan PCB", tag, owner, clock, "nes_waixing_sgzlz", __FILE__)
```
Comment (cuavas) [inline src/devices/bus/nes/waixing.cpp]: You can initialise the array here using C++11 uniform initialisation syntax.

File: src/devices/bus/nes/waixing.cpp [hunk 4]
```diff
@@ -1035,30 +1035,69 @@ WRITE8_MEMBER(nes_waixing_sgz_device::write_h)
...
+		// case 0x701:
+		//  reg1 = (value >> 1);
+		//  bank = reg1 + (reg2 << 2);
+		//  prg32(bank);
+		//  break;
+		// case 0x702:
+		//  reg2 = value;
+		//  break;
+
+		if (m_reg[0] & 2)
```
Comment (cuavas) [inline src/devices/bus/nes/waixing.cpp]: Please use the `BIT` helper for these kinds of tests.

File: src/devices/bus/nes_ctrl/suborkey.cpp [hunk 5]
```diff
@@ -40,10 +40,10 @@ static INPUT_PORTS_START( fc_suborkey )
 	PORT_START("SUBOR.2")
 	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_INSERT) PORT_CHAR(UCHAR_MAMEKEY(INSERT))
 	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_BACKSPACE) PORT_CHAR(8)
-	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("NEXT")
+	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_PGDN) PORT_CHAR(UCHAR_MAMEKEY(PGDN))
```
Comment (cuavas) [inline src/devices/bus/nes_ctrl/suborkey.cpp]: If the key is labelled "NEXT" that should be preserved here.

File: src/devices/bus/nes_ctrl/suborkey.cpp [hunk 6]
```diff
@@ -40,10 +40,10 @@ static INPUT_PORTS_START( fc_suborkey )
...
 	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_RIGHT) PORT_CHAR(UCHAR_MAMEKEY(RIGHT))
 	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_F8) PORT_CHAR(UCHAR_MAMEKEY(F8))
-	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_NAME("PRIOR")
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD ) PORT_CODE(KEYCODE_PGUP) PORT_CHAR(UCHAR_MAMEKEY(PGUP))
```
Comment (cuavas) [inline src/devices/bus/nes_ctrl/suborkey.cpp]: Same here - the keycap label should be preserved.

**Comment by cuavas** [comment]
Also, please fix your git user information - you need to put an e-mail address in.  You can use the @users.noreply.github.com one if you don't want to dox yourself.

**Comment by kazblox** [comment]
There, just done how the way you like it. If there's any more problems with these requested changes, let me know.

**Pull request merged**